src/bin/server/stream.rs

```diff
@@ -0,0 +1,209 @@
+                .streaming(ChunkedStreamReader::new(
+                    self.reader,
+                    HttpRange {
+                        start: 0,
+                        length: self.size,
+                    },
+                    req.cpu_pool().clone(),
+                )),
+            Err(HttpRangeError::InvalidRange) => HttpResponse::RangeNotSatisfiable().finish(),
+            Err(HttpRangeError::NoOverlap) => {
+                HttpResponse::build(StatusCode::RANGE_NOT_SATISFIABLE)
+                    .header(header::CONTENT_RANGE, format!("bytes */{}", self.size))
+                    .finish()
+            }
+        };
+
+        Ok(response)
+    }
+}
+
+/// A stream which reads from an io.Read + io.Seek in chunks.
+struct ChunkedStreamReader<R>
+where
+    R: RandomRead + Send + 'static,
+{
+    /// A container for the stream. This is used to manually keep track of the lifetime of the
+    /// reader.
+    stream_option: Option<R>,
+
+    /// A future which resolves once a chunk is read.
+    future: Option<CpuFuture<(R, Bytes), io::Error>>,
+
+    /// The pool the task of reading from the reader will be executed on.
+    cpu_pool: CpuPool,
+
+    /// When the offset is greater than or equal to this value, stop reading from the stream.
+    upper_bound: u64,
+
+    /// The number of bytes from the beginning of the file the next read should start from.
+    offset: u64,
+
+    /// The maximum size of the size of each chunk. The last chunk may be shorter.
+    chunk_size: u64,
+}
+
+impl<R> ChunkedStreamReader<R>
+where
+    R: RandomRead + Send + 'static,
+{
+    fn new(stream: R, range: HttpRange, cpu_pool: CpuPool) -> ChunkedStreamReader<R> {
+        ChunkedStreamReader {
+            stream_option: Some(stream),
+            future: None,
+            cpu_pool,
+            upper_bound: range.start + range.length,
+            offset: range.start,
+            chunk_size: u16::max_value() as u64,
+        }
+    }
+
+    /// Waits for the current future chunk of bytes to resolve, then returns it. Panics if there is
+    /// no future set.
+    fn wait_for_future(&mut self) -> Poll<Option<Bytes>, io::Error> {
+        let (stream, bytes) = try_ready!(self.future.as_mut().expect("no future set").poll());
+
+        self.future.take();
+        self.stream_option = Some(stream);
+        self.offset += bytes.len() as u64;
+
+        Ok(Async::Ready(Some(bytes)))
+    }
+}
+
+impl<R> Stream for ChunkedStreamReader<R>
+where
+    R: RandomRead + Send + 'static,
+{
+    type Item = Bytes;
+    type Error = io::Error;
+
+    fn poll(&mut self) -> Poll<Option<Bytes>, io::Error> {
+        if self.future.is_some() {
+            return self.wait_for_future();
+        }
+
+        if self.offset >= self.upper_bound {
+            // End of stream.
+            return Ok(Async::Ready(None));
+        }
+
+        let upper_bound = self.upper_bound;
+        let offset = self.offset;
+        let chunk_size = self.chunk_size;
```

Comment:
Self can't be captured by the move closure because the reference to self isn't valid for the static lifetime. Since self can't be captured, none of its properties can be accessed either.